flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java

```diff
@@ -74,22 +79,46 @@
 			@Nonnull ResourceManager<?> resourceManager,
 			@Nonnull LeaderRetrievalService dispatcherLeaderRetrievalService,
 			@Nonnull LeaderRetrievalService resourceManagerRetrievalService,
-			@Nonnull WebMonitorEndpoint<?> webMonitorEndpoint) {
+			@Nonnull WebMonitorEndpoint<?> webMonitorEndpoint,
+			@Nonnull FatalErrorHandler fatalErrorHandler,
+			@Nonnull CompletableFuture<DispatcherGateway> dispatcherGatewayCompletableFuture) {
 		this.dispatcherRunner = dispatcherRunner;
 		this.resourceManager = resourceManager;
 		this.dispatcherLeaderRetrievalService = dispatcherLeaderRetrievalService;
 		this.resourceManagerRetrievalService = resourceManagerRetrievalService;
 		this.webMonitorEndpoint = webMonitorEndpoint;
+		this.fatalErrorHandler = fatalErrorHandler;
 		this.terminationFuture = new CompletableFuture<>();
 		this.shutDownFuture = new CompletableFuture<>();
 
 		registerShutDownFuture();
+		failOnPrematureTermination(dispatcherGatewayCompletableFuture);
 	}
 
 	private void registerShutDownFuture() {
 		FutureUtils.forward(dispatcherRunner.getShutDownFuture(), shutDownFuture);
 	}
 
+	private void failOnPrematureTermination(CompletableFuture<DispatcherGateway> dispatcherGatewayCompletableFuture) {
+		dispatcherGatewayCompletableFuture.whenComplete((dispatcher, throwable) -> {
+			if (dispatcher != null && dispatcher instanceof Dispatcher) {
+				CompletableFuture.anyOf(((Dispatcher) dispatcher).getTerminationFuture(), resourceManager.getTerminationFuture())
+					.whenComplete((ignored, t) -> {
+						if (isRunning.get()) {
+							fatalErrorHandler.onFatalError(t);
```

Comment:
It would be good to log a warning here that the DRMC shut down because the dispatcher/rm unexpected terminated.